From arjan@infradead.org  Tue Sep 23 15:53:52 2008
From: Arjan van de Ven <arjan@linux.intel.com>
Date: Sat, 20 Sep 2008 19:08:39 -0700
Subject: debug: Introduce a dev_WARN() function
To: Greg KH <greg@kroah.com>
Message-ID: <20080920190839.720b9602@infradead.org>


From: Arjan van de Ven <arjan@linux.intel.com>

in the line of dev_printk(), this patch introduces a dev_WARN() function,
that takes a struct device and then a printk format/args set of arguments.
Unlike dev_printk(), the effect is that of WARN() in that a full warning
message (including filename/line, module list, versions and a backtrace)
is printed in addition to the device name and the arguments.

Signed-off-by: Arjan van de Ven <arjan@linux.intel.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 include/linux/device.h |    8 ++++++++
 1 file changed, 8 insertions(+)

--- a/include/linux/device.h
+++ b/include/linux/device.h
@@ -570,6 +570,14 @@ extern const char *dev_driver_string(con
 	({ if (0) dev_printk(KERN_DEBUG, dev, format, ##arg); 0; })
 #endif
 
+/*
+ * dev_WARN() acts like dev_printk(), but with the key difference
+ * of using a WARN/WARN_ON to get the message out, including the
+ * file/line information and a backtrace.
+ */
+#define dev_WARN(dev, format, arg...) \
+	WARN(1, "Device: %s\n" format, dev_driver_string(dev), ## arg);
+
 /* Create alias, so I can be autoloaded. */
 #define MODULE_ALIAS_CHARDEV(major,minor) \
 	MODULE_ALIAS("char-major-" __stringify(major) "-" __stringify(minor))
